### ISSUE TITLE
Build a mechanism for notifications on the web site, not using email

### ISSUE THREAD

**Issue by woodpeck** [openstreetmap/openstreetmap-website#1387]
The following request has been made on help.openstreetmap.org by someone who did not dare add it here on GithHub directly:

> My specific concern is that users should be able to receive notifications of changeset discussions on the OSM website itself, in addition to, or instead of, being limited to just email notifications.

Source: https://help.openstreetmap.org/questions/53264/where-to-make-a-feature-request-concerning-changeset-comment-notifications

**Comment by tomhughes** [comment]
Actually he said that he couldn't understand github, not that he didn't dare raise it here... So it might have made more sense to just point at the right page to add an issue...

**Comment by simonpoole** [comment]
https://github.com/openstreetmap/openstreetmap-website/issues/908 should be referenced here. Instead of sending a message, maybe we should simply have an API that returns the number of new changeset comments since the last "login". This would be enough to allow editor devs to show a message/popup whatever.

**Comment by Paul012** [comment]
Thanks, Woodpeck, for helping post this, and Simonpoole for the link. Seeing as the earlier issue pretty adequately addresses my concerns, I'd be fine if this was closed as a duplicate.

**Comment by boothym** [comment]
This came up recently when a user said emails were going to his spam in gmail, meaning many changeset comments went unanswered and only a user block got them to communicate. They complained that there are other ways of contacting them, which I suppose is true. 

If you send someone a message, they received a [x] notification in the top corner of the website, and also JOSM alerts you on startup when you have an unread message. The same needs to be done for changeset comments.

**Comment by aleung** [comment]
This issue is inactive for a time. Adding a notification on website, just like what GitHub has for issue/PR, would be helpful. 

Nowaday email is partial replaced by mobile message app. I even not check my mail for week.

**Comment by aawiseman** [comment]
Yes, agreed. I think other ways to make it clear you got a message or comment would help.

**Comment by Cj-Malone** [comment]
+2 for notifications to be shown on the website, and made available via an API like the message notifications.

I think for most people email isn't manageable any more, so having it be the only place changeset comments go isn't ideal, I don't think we can judge anyone for not replying to a changeset comment until something like this is implemented.

**Comment by Dimitar5555** [comment]
I also think that notifications should be (at the very least) shown on the website. I'm copying my suggestion for implementing it from the mentioned issue.
> It can be based on messages where if the sender_id is 0, it can be hidden from the messages inbox and shown in the notification's inbox. That way you won't have to add new tables to the DB. Notifications can be received for the same things that user's receive emails (notes, changeset comments, somebody friended them etc.).

Another option would be to create a clone of the messages table but without `from_user_id` and `from_user_visible`. Potentially `body_format`, `to_user_visible` and `body` could also be omitted for notifications if they are not needed.

**Comment by kcne** [comment]
After a bit of research on some libraries that offer such functionalities and implementation details that we could potentially use for this:
### 1. [notifications-rails](https://github.com/jonhue/notifications-rails)
`notifications-rails` provides a modular structure for creating, rendering, pushing, and managing settings for notifications, with each module backed by the database. This setup supports on-site notifications with user-configurable preferences, allowing users to see unread counts and mark notifications as read.

**Database Structure and Requirements:**

- **New Tables:**
  - **`notifications`:** Main table storing each notification. Key columns:
    - `target_id` and `target_type` (polymorphic association for the recipient)
    - `object_id` and `object_type` (polymorphic link to the notification’s subject, e.g., a `Comment`)
    - `metadata` (JSON for additional data), `category` (notification type), and `read` (boolean for read/unread).
  - **`notification_groups` (optional):** Allows bulk notifications for groups (e.g., subscribers).

- **Modified Tables:**
  - **User Table**: Adds optional columns for `read_notification_count` and `unread_notification_count` to cache unread counts, and `settings` for storing user-specific notification preferences.

- **Indexes:**  
  - Indexes on `target_id`, `target_type`, and `category` for quick lookups in the `notifications` table.

**Database Workflow:**
1. **Creating Notifications:** Notifications are inserted into `notifications` with details on target, subject, metadata, and category.
2. **Marking as Read:** Updating `read` marks notifications as read, and `unread_notification_count` can be adjusted for caching.
3. **Bulk Notifications:** The `notification_groups` table can manage batch notifications by defining recipient groups.

### 2. [noticed](https://github.com/excid3/noticed)

`noticed` offers individual and bulk notifications, supporting multiple delivery methods, such as ActionCable for real-time updates. It’s suitable for creating on-site notifications tied to specific events, each managed through ActiveJob for asynchronous processing.

**Database Structure and Requirements:**

- **New Tables:**
  - **`notified_events`:** Logs each event that triggers notifications. Key columns:
    - `type` (event type, e.g., “NewCommentNotifier”)
    - `record_id` and `record_type` (polymorphic association to link with the event’s subject, e.g., `Post`)
    - `params` (JSON to store event details, like message and URL).
  
  - **`notifications`:** Links notifications to events and tracks each recipient. Key columns:
    - `event_id` (foreign key to `notified_events`), `recipient_id` and `recipient_type` (polymorphic association for the recipient), and `read_at` (datetime for read status).

- **Modified Tables:**
  - **User Table**: No required modifications, but adding `has_many :notifications` makes it easier to manage notifications.

- **Indexes:**
  - Indexes on `type`, `record_type`, and `record_id` in `notified_events` for efficient filtering and quick retrieval.

**Database Workflow:**
1. **Logging Events:** Each new event (e.g., a comment) is logged in `notified_events`, capturing relevant metadata.
2. **Creating Notifications:** Notifications are linked to events in `notifications`, enabling individual tracking for each recipient.
3. **Read Status:** Users mark notifications as read by updating `read_at`, enabling quick querying for unread notifications and visual badges.

Hope you find this helpful @gravitystorm

**Comment by pablobm** [comment]
Hello @kcne 👋  I was having a look at your proposals. My impression is that [noticed](https://github.com/excid3/noticed) is the better option as:
- It's an active project, with a responsive maintainer.
- It's widely used (I asked around a bit and it seems to be the preferred option).
- The design seems to use the right abstraction, separating the events from the multiple notifications they may generate (in line with @gravitystorm's thoughts at https://github.com/openstreetmap/openstreetmap-website/issues/908#issuecomment-508106079).

Is this something you would still want (and have time) to work on?

**Comment by kcne** [comment]
Hey Pablo,  
Thanks for the thoughtful review I've been following your activity for a while but was caught up with work(keep up the good work). I'm definitely interested in working on this at my own pace. I'll keep you posted here and open a draft PR once I have something ready to share.

Thanks.

**Comment by pablobm** [comment]
Amazing! Here's an idea: would you be able to work on a draft PR from the start?

My thinking is to have a way to keep track on progress as you go, and I can review and test early, using the PR for communication. I don't want to rush you (thank you for volunteering!) but also I want to avoid a situation where your circumstances change, you don't have time any more, and we lose track of progress. Does that make sense?

And of course: if you find something where others can help, please let us know! I'll be around and part of my job is to make it easier for contributors. I will be keeping an eye anyway, and will let you know if there are any changes that could affect your work.
